broker.go

```diff
@@ -853,6 +881,124 @@ func (b *Broker) sendAndReceiveSASLPlainAuth() error {
 	return nil
 }
 
+// sendAndReceiveSASLOAuth performs the authentication flow as described by KIP-255
+// https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=75968876
+func (b *Broker) sendAndReceiveSASLOAuth(tokenProvider OAuthBearerTokenProvider) error {
+
+	// This version allows us to wrap tokens in the Kafka protocol, as opposed
+	// to sending opaque packets
+	handshakeVersion := int16(1)
+
+	if err := b.sendAndReceiveSASLHandshake(SaslTypeOAuth, handshakeVersion); err != nil {
+		Logger.Printf("Error while performing SASL handshake %s\n", b.addr)
+		return err
+	}
+
+	requestTime := time.Now()
+
+	var bytesWritten int
+	var err error
+
+	if bytesWritten, err = b.sendSASLOAuthBearerClientResponse(tokenProvider.Token()); err != nil {
+		return err
+	}
+
+	b.updateOutgoingCommunicationMetrics(bytesWritten)
+
+	var bytesRead int
+
+	if bytesRead, err = b.receiveSASLOAuthBearerServerResponse(); err != nil {
+		return err
+	}
+
+	requestLatency := time.Since(requestTime)
+	b.updateIncomingCommunicationMetrics(bytesRead, requestLatency)
+
+	return nil
+}
+
+func (b *Broker) sendSASLOAuthBearerClientResponse(bearerToken []byte) (int, error) {
+
+	// Initial client response as described by RFC-7628
+	// https://tools.ietf.org/html/rfc7628
+	oauthRequest := []byte(`n,,`)
+	oauthRequest = append(oauthRequest, '\x01')
+	oauthRequest = append(oauthRequest, []byte(`auth=Bearer `)...)
+	oauthRequest = append(oauthRequest, bearerToken...)
+	oauthRequest = append(oauthRequest, '\x01')
+	oauthRequest = append(oauthRequest, '\x01')
+
+	rb := &SaslAuthenticateRequest{oauthRequest}
+
+	req := &request{correlationID: b.correlationID, clientID: b.conf.ClientID, body: rb}
+
+	var buf []byte
+	var err error
+
+	buf, err = encode(req, b.conf.MetricRegistry)
+
+	if err != nil {
+		return 0, err
+	}
+
+	if err := b.conn.SetWriteDeadline(time.Now().Add(b.conf.Net.WriteTimeout)); err != nil {
+		return 0, err
+	}
+
+	var bytesWritten int
+
+	if bytesWritten, err = b.conn.Write(buf); err != nil {
+		Logger.Printf("Failed to send SASL/OAUTHBEARER initial request %s: %s\n", b.addr, err.Error())
+		return 0, err
+	}
+
+	b.correlationID++
+
+	return bytesWritten, nil
+}
+
+func (b *Broker) receiveSASLOAuthBearerServerResponse() (int, error) {
+
+	var bytesRead int
+	var err error
+	totalBytesRead := 0
+
+	header := make([]byte, 8)
+
+	if bytesRead, err = io.ReadFull(b.conn, header); err != nil {
+		Logger.Printf("Failed to read SASL/OAUTHBEARER intitial response header : %s\n", err.Error())
```

Comment:
as suggested above, don't log if you are returning error, these errors have potential of filling the disks if not handled properly downstream.

Reply:
Makes sense. How about further down where the authentication failure is logged (`Invalid SASL/OAUTHBEARER request`)? This would help users figure out why their client doesn't work when they pass invalid credentials.

Reply:
I could make the caller log the auth message. Not sure if that's a disk space issue if the error is only emitted when debug mode is turned on?